nncf/openvino/graph/model_builder.py

```diff
@@ -0,0 +1,222 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import deque
+from typing import Dict, List, Tuple
+
+import openvino.runtime as ov
+from openvino.runtime import opset13 as opset
+from openvino.runtime.utils.node_factory import NodeFactory
+
+from nncf.openvino.graph.model_transformer import OVModelTransformer
+from nncf.openvino.graph.node_utils import get_parameter_node_name
+from nncf.openvino.graph.node_utils import get_result_node_name
+
+
+class ModelBuilder:
+    """
+    The purpose of the ModelBuilder is to build a new OpenVINO model from input and output points.
+    This Builder was created to reduce the number of model cloning that is required for ModelTransformer to work.
+    """
+
+    def __init__(self):
+        self._node_factory = NodeFactory()
+
+    @staticmethod
+    def _create_parameter(node_name: str, node_input: ov.Input) -> ov.Node:
+        """
+        A method that contains steps to create a Parameter for a new model using a specific template.
+        """
+        port_id = node_input.get_index()
+        parameter_name = get_parameter_node_name(node_name, port_id)
+        return opset.parameter(
+            shape=node_input.get_partial_shape(),
+            dtype=node_input.get_element_type(),
+            name=parameter_name,
+        )
+
+    @staticmethod
+    def _create_result(node_name: str, node_output: ov.Input) -> ov.Node:
+        """
+        A method that contains steps to create a Result for a new model using a specific template.
+        """
+        port_id = node_output.get_index()
+        result_name = get_result_node_name(node_name, port_id=port_id)
+        result = opset.result(node_output, name=result_name)
+        result.get_output_tensor(0).set_names({result_name})
+        return result
+
+    def _collect_graph_nodes(
+        self,
+        input_ids: List[Tuple[str, int]],
+        output_ids: List[Tuple[str, int]],
+        node_mapping: Dict[str, ov.Node],
+    ) -> List[ov.Node]:
+        """
+        A method for aggregating layers to be further cloned.
+        Aggregation is designed in such a way that layers are listed from right to left,
+        as they pass from bottom to top. This is done in order to find all constants in the model and
+        to start graph creation from them (as well as Parameter layers), because
+        OpenVINO graph is created from top-down and cannot be created otherwise.
+
+        Legend: w - weigths, c - convert, il/lh - input low/high, ol/oh - output low/high
+        (w)
+         |
+        (c) (il) (ih) (ol) (oh)
+          \   |   |   /   /
+           (fake quantize) (parameter)
+                  \           /
+                  (convolution)
+                        |
+                    (result)
+        Based on the above graph, the return value would look like this:
```

Comment:
What are `input_ids` and `output_ids` in that example? Or you just want to demonstrate how nodes are organized inside returned list?

Reply:
Basically, `input_ids` and `output_ids` are the lists of the points in the special format - `(node_name, port_id)`. This helps to point to the precise part of the model that may be used to define the subgraph.
I'll add this to the description.